### PR TITLE
Feature/issue 1951 compound declaration definition

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submisison Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary

Adds compound declaration/definition statements such as `real x = 1.0` to Stan language.

#### Intended Effect

For any variable that is assignable, allow assignment to be combined with the variable definition.
Assignable variable are local variables and variables declared in the transformed data, transformed parameters, and generated quantities blocks.

#### How to Verify

To verify that the parser correctly recognizes valid compound declaration/definition statements, run unit test `stan/src/test/unit/lang/parser/var_decls_grammar_test.cpp`.

To verify that the generated code compiles correctly, make the test models in `src/test/test-models/good/declare-define-var*.stan`,  e.g.:
```
for x in src/test/test-models/good/declare-define-*.stan
do
echo ${x%.stan}
make ${x%.stan}.hpp-test
done
```

#### Side Effects

The compiler cannot check that the dimension sizes match.  See example model: `src/test/test-models/good/declare-define-var-vec-runtime-error.stan`

#### Documentation

Basic syntax described in Stan manual.  More documentation is needed.  Future planned work is to update the example models to use this feature.

#### Reviewer Suggestions

Daniel Lee

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Mitzi Morris

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: src/test/test-models/good/declare-define-var-double.stan [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  real tp4[5];
+  real tp5[5] = tp4;
+  tp3 = tp1 + tp2;
+
+  // local variables
+  {
+    real lp1 = 1;      // real_p <- int_d 
+    real lp2 = 2.0;    // real_p <- real_d
+    real lp3 = tp2;      // real_p <- real_p
+      real lp4[6];
```
Comment (syclik) [inline src/test/test-models/good/declare-define-var-double.stan]: fix indenting

**Comment by bgoodri** [comment]
I still think the parser should warn if the user declares without defining on the same line in all blocks except data and parameters.

File: src/docs/stan-reference/language.tex [hunk 2]
```diff
@@ -1267,6 +1267,84 @@ \subsection{Type Naming Notation}
...
+even though \code{0} is of type \code{int} and \code{sum} is of type
+\code{real}, because integer-typed scalar expressions can be assigned
+to real-valued scalar variables.  In all other cases, the type of the
+expression on the right-hand side of the assignment must be identical
+to the type of the variable being declared.
+
+Any type may be assigned.  For example,
+%
+\begin{stancode}
+real matrix[3, 2] a = b;
```
Comment (mitzimorris) [inline src/docs/stan-reference/language.tex]: changes made and pushed

File: src/docs/stan-reference/language.tex [hunk 3]
```diff
@@ -1267,6 +1267,84 @@ \subsection{Type Naming Notation}
...
+their static typing and cannot be validated until run time.
+
+
+\subsection{Right-Hand Side Expressions}
+
+The right-hand side may be any expression which has a type which is
+assignable to the variable being declared.  For example,
+%
+\begin{stancode}
+real matrix[3, 2] a = 0.5 * (b + c);
```
Comment (mitzimorris) [inline src/docs/stan-reference/language.tex]: changes made and pushed

File: src/test/test-models/good/declare-define-var-double.stan (hunk 1 above)
Comment (mitzimorris) [inline src/test/test-models/good/declare-define-var-double.stan]: changes made and pushed

**Comment by mitzimorris** [comment]
would prefer to create new issue for adding parser warning for variables declarations with missing definitions in blocks which allow compound declaration/definition statements.

**Comment by syclik** [comment]
+1.

**Comment by bob-carpenter** [comment]
Because the only way to define a variable is through assignment
(e.g., no pass by reference), then it's relatively easy to
figure out if a variable's being used before it's defined, which
is really where we want the warning to go.  It'd be easy to put
one in a declaration without a definition, but it might be confusing
to users who've actually defined the variable.

**Comment by bgoodri** [comment]
One of the more common problems with users' Stan programs is

model {
  real z;
  z ~ normal(0, 1);
}

so nothing actually gets assigned to z. I presumed that since we didn't
already have a parser warning for that case, it was difficult to determine
whether to throw it. Warning whenever z is declared but not defined would
fix that, although it would be somewhat overkill for

model {
  real z;
  z = 0;
}

but there is nothing particularly to be said in favor of writing a model
block like that, other than it has always been mandatory.

**Comment by bob-carpenter** [comment]
> On Oct 4, 2016, at 10:47 PM, bgoodri <notifications@github.com> wrote:
> 
> One of the more common problems with users' Stan programs is
> 
> model {
> real z;
> z ~ normal(0, 1);
> }

Agreed that this should produce a warning.  And I think the
warning should point to where the variable is used before being
defined, that is in the ~ statement. 

Harder to detect case:

  real z[10];

  for (n in 1:9)
    z[n] = ... something the user couldn't vectorize ... 

Can't do this statically.

> so nothing actually gets assigned to z. I presumed that since we didn't
> already have a parser warning for that case, it was difficult to determine
> whether to throw it. Warning whenever z is declared but not defined would
> fix that, although it would be somewhat overkill for
> 
> model {
> real z;
> z = 0;
> }

This one should not produce a warning.

I'd be OK producing warnings in cases where you have a declaration without
a definition if we had a way to turn those warnings off.  That's #1978 and
#1672 and is related to #512 and is also related to the "pedantic" mode
proposal:  https://github.com/stan-dev/stan/wiki/Stan-Parser-Pedantic-Mode

- Bob

**Pull request merged**